src/abe/dippe.c

```diff
@@ -0,0 +1,613 @@
+    cfe_vec_init(pol, (strlen(pattern)+1));
+    mpz_set_ui(pol->vec[strlen(pattern)], 0);
+    for (unsigned int i=0; i<strlen(pattern); i++) {
+        if (pattern[i] == '1') {
+            cfe_uniform_sample(pol->vec[i], dippe->p);
+            mpz_sub(pol->vec[strlen(pattern)], pol->vec[strlen(pattern)], pol->vec[i]);
+        } else if (pattern[i] == '0'){
+            mpz_set_ui(pol->vec[i], 0);
+        }
+        else return CFE_ERR_MALFORMED_INPUT;
+    }
+    mpz_mod(pol->vec[strlen(pattern)], pol->vec[strlen(pattern)], dippe->p);
+    return CFE_ERR_NONE;
+}
+
+/**
+ * Helper function that build a conjunction attribute vector out of a given pattern
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param attrs A pointer to a mpz_t struct; Represents the resulting attribute vector
+ * @param pattern String that is used as template for the attribute vector
+ * @return Error code
+ */
+cfe_error cfe_dippe_build_conjunction_attribute_vector(cfe_dippe *dippe, mpz_t *attrs, const char pattern[]) {
+    mpz_init((*attrs));
+    mpz_set_ui((*attrs), 0);
+    for (unsigned int i=0; i<strlen(pattern); i++) {
+        if (pattern[i] == '1') {
+           mpz_setbit((*attrs), i);
+        }
+        else if (pattern[i] != '0'){
+            return CFE_ERR_MALFORMED_INPUT;
+        }
+    }
+    mpz_setbit((*attrs), strlen(pattern));
+    return CFE_ERR_NONE;
+}
+
+/**
+ * Helper function that build an exact threshold policy vector out of a given pattern
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param pol A pointer to a cfe_vec struct; Represents the resulting policy vector
+ * @param pattern String that is used as template for the policy vector
+ * @param threshold Threshold value
+ * @return Error code
+ */
+cfe_error cfe_dippe_build_exact_threshold_policy_vector(cfe_dippe *dippe, cfe_vec *pol, const char pattern[], unsigned long threshold) {
+    cfe_vec_init(pol, (strlen(pattern)+1));
+    mpz_set_ui(pol->vec[strlen(pattern)], threshold);
+    for (unsigned int i=0; i<strlen(pattern); i++) {
+        if (pattern[i] == '1') {
+            mpz_set_ui(pol->vec[i], 1);
+        } else if (pattern[i] == '0'){
+            mpz_set_ui(pol->vec[i], 0);
+        }
+        else return CFE_ERR_MALFORMED_INPUT;
+    }
+    mpz_neg(pol->vec[strlen(pattern)], pol->vec[strlen(pattern)]);
+    mpz_mod(pol->vec[strlen(pattern)], pol->vec[strlen(pattern)], dippe->p);
+    return CFE_ERR_NONE;
+}
+
+/**
+ * Helper function that build an exact threshold attribute vector out of a given pattern
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param attrs A pointer to a cfe_dippe_vec struct; Represents the resulting attribute vector
+ * @param pattern String that is used as template for the attribute vector
+ * @return Error code
+ */
+cfe_error cfe_dippe_build_exact_threshold_attribute_vector(cfe_dippe *dippe, mpz_t *attrs, const char pattern[]) {
```

Comment:
Isn't this just copy paste of `cfe_dippe_build_conjunction_attribute_vector`?

Reply:
Correct, I removed one of them.